runtime/shared_heap.c

```diff
@@ -785,6 +811,475 @@ void caml_verify_heap(caml_domain_state *domain) {
+  For the first phase we need not consider full pools, they
+  cannot be evacuated to or from. */
+  caml_global_barrier();
+
+  struct caml_heap_state* heap = Caml_state->shared_heap;
+
+  #ifdef DEBUG
+  /* Check preconditions for the heap: */
+  for(int sz_class = 1; sz_class < NUM_SIZECLASSES; sz_class++) {
+    /* No sweeping has happened yet */
+    CAMLassert(heap->avail_pools[sz_class] == NULL);
+    CAMLassert(heap->full_pools[sz_class] == NULL);
+    CAMLassert(heap->swept_large == NULL);
+    /* No pools waiting for adoption */
+    if( participants[0] == Caml_state ) {
+      CAMLassert(pool_freelist.global_avail_pools[sz_class] == NULL);
+      CAMLassert(pool_freelist.global_full_pools[sz_class] == NULL);
+    }
+    /* The minor heap is empty */
+    CAMLassert(Caml_state->young_ptr == Caml_state->young_end);
+    /* The mark stack is empty */
+    CAMLassert(caml_mark_stack_is_empty());
+  }
+  #endif
+
+  for(int sz_class = 1; sz_class < NUM_SIZECLASSES; sz_class++) {
+    /* We only care about moving things in pools that aren't full (we cannot
+    evacuate to or from a full pool) */
+    pool* cur_pool = heap->unswept_avail_pools[sz_class];
+    /* [evac_pool] will eventually point to the first pool in this size class
+      that is to be evacuated */
+    pool* evac_pool = NULL;
+    int nonempty_blocks = 0;
+
+    if( cur_pool == NULL ) {
+      /* No partially filled pools for this size, nothing to do */
+      continue;
+    }
+
+    /* We need to calculate the number of non-empty blocks in the pools.
+       note: this is an over-approximation of the amount of space we'll need.
+       We only move _live_ blocks from evacuated pools (the from space)
+       in to free space in the allocating pools (the to space). There will be
+       garbage in the allocating pools that isn't cleaned up and which we
+       can't use as free space yet. */
+    while( cur_pool != NULL ) {
+      value* p = (value*)((char*)cur_pool + POOL_HEADER_SZ);
+      value* end = (value*)cur_pool + POOL_WSIZE;
+      mlsize_t wh = wsize_sizeclass[sz_class];
+
+      while (p + wh <= end) {
+        header_t hd = (header_t)atomic_load_relaxed((atomic_uintnat*)p);
+
+        /* Because we've rotated the colours, things we marked in the previous
+           cycle is now UNMARKED */
+        if( hd != 0 ) {
+          nonempty_blocks += 1;
+        }
+
+        p += wh;
+      }
+
+      cur_pool = cur_pool->next;
+    }
+
+    if( nonempty_blocks == 0 ) {
+      /* No non-empty blocks in partially filled pools, nothing to do */
+      continue;
+    }
+
+    /* cur_pool is now the pool we're evacuating (allocating) to */
+    cur_pool = heap->unswept_avail_pools[sz_class];
+    /* evac_pool will be the pool we're evacuating from */
+
+    int cur_blocks = 0;
+
+    /* Now decide which pools will be allocated in to and which will be
+      evacuated. */
+    while( cur_pool != NULL ) {
+      if( cur_blocks > nonempty_blocks ) {
```

Comment:
Should be `>=` (although this is not an error, it will mean that we don't fill up the final non-evacuating block).